### DISCUSSION TITLE
Use Query is always loading?

### DISCUSSION THREAD

**Discussion by danieludokike** [reactive-python/reactpy-django#239] [Problem] [answered]
I have followed the instructions in the doc and can't pull data from the django db using the use_query. What am I doing wrongly?
![img](https://github.com/reactive-python/reactpy-django/assets/68450806/36a09d56-3af8-4932-9ae1-fbd44bcee1f3)
![img2](https://github.com/reactive-python/reactpy-django/assets/68450806/83acee8c-df97-437f-99c4-e6003b00d363)
<img width="466" alt="img3" src="https://github.com/reactive-python/reactpy-django/assets/68450806/bdd4c8cd-1868-4574-8786-412451a1fc37">
![img4](https://github.com/reactive-python/reactpy-django/assets/68450806/62fae563-f531-400d-b4a0-c55696f93ae2)

**Comment by Archmonger** [comment]
Are you getting any errors in your browser console?
Are you getting any errors in your python terminal?
What version of Django and Reactpy-Django are you using?
Did you make sure to run database migrations prior to running this example?

**Reply by danieludokike** [reply]
I am using the latest version of django and reactpy-django. The browser and terminal has no issue and I made migrations and migrated as required. Can't tell where the issue might be coming from. Though I had the same issue @ 3.8.0 release when I knwe about the the library and wanted to explore, I stopped because of the issue and still facing now that I want to build an SPA with it. Please help. I have attachments that can assist below.

[requirements.txt](https://github.com/user-attachments/files/15882834/requirements.txt)
<img width="741" alt="console_output" src="https://github.com/reactive-python/reactpy-django/assets/68450806/f4462eb9-189a-47b7-a303-c12ad9e591c9">
<img width="482" alt="migrations_confirmed" src="https://github.com/reactive-python/reactpy-django/assets/68450806/fee39f33-34f2-4f82-bdf5-d6d972506abc">
<img width="533" alt="terminal_output" src="https://github.com/reactive-python/reactpy-django/assets/68450806/e96f0c0f-f9cf-4d82-b011-ebb8d226989d">

**Reply by danieludokike** [reply]
I attached two more images to confirm that the Model migrations were
![add_todo_in_admin](https://github.com/reactive-python/reactpy-django/assets/68450806/e6b2c08e-4da9-4997-a0df-92d579659bd7)
![Screenshot_18-6-2024_92818_127 0 0 1](https://github.com/reactive-python/reactpy-django/assets/68450806/3ec14c22-5bdd-4478-8154-4625c5eaf18a)
 as successful as expected.

**Reply by Archmonger** [reply, marked as answer]
There was a typo in the docs, thank you for noticing this. [I will fix the docs ASAP](https://github.com/reactive-python/reactpy-django/pull/240).

```py
rendered_items = html.ul([html.li(item, key=item) for item in item_query.data])
```
needs to be changed to
```py
rendered_items = html.ul([html.li(item, key=item.pk) for item in item_query.data])
```

The reason is `key=...` must always be a unique [immutable](https://realpython.com/python-mutable-vs-immutable-types/#immutable-built-in-data-types-in-python) value.

The value `item.pk` is both a string and unique.

**Reply by danieludokike** [reply]
Great!

**Reply by danieludokike** [reply]
The complete docs could be
rendered_items = html.ul([html.li(item.text, key=item.pk) for item in item_query.data])

**Reply by danieludokike** [reply]
Again, is there a way to subscript queries? Traditional django could be TodoItem.objects.all()[:10] any way to do that using the sync_to_async approach?

**Reply by Archmonger** [reply]
I'm pretty sure `database_sync_to_async` supports slices:
```python
return await database_sync_to_async(TodoItem.objects.all)()[:10]
```

Async within the Django ORM is hit or miss though. If that throws an error, then you'll need to create a wrapper function.

```python
def django_orm_query():
    return TodoItem.objects.all()[:10]

async def get_items():
    return await database_sync_to_async(django_orm_query)()
```

But ReactPy automatically does this kind of wrapper for sync queries, so for convenience you can write your query as a sync function.

```python
def get_items():
    return TodoItem.objects.all()[:10]
```

**Reply by danieludokike** [reply]
`return await database_sync_to_async(TodoItem.objects.all)()[:10]` Slicing directly produces a coroutine not subscriptable error
![err](https://github.com/reactive-python/reactpy-django/assets/68450806/01ebeba0-7dbe-466e-8f43-a30939b3fefc)


Like wise `def get_items():
    return TodoItem.objects.all()[:10]` produces error too.  
![err2](https://github.com/reactive-python/reactpy-django/assets/68450806/c713e284-cbd1-43ed-896d-ac7507aaf5c9)


What worked perfectly was  `def django_orm_query():
    return TodoItem.objects.all()[:10]

async def get_items():
    return await database_sync_to_async(django_orm_query)()`
